keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md

```diff
@@ -382,57 +373,57 @@ It may be defined by command line option `--horizontal-pod-autoscaler-downscale-
 
 The following API changes are needed:
 
-We should add `Scale{Up,Down}{Percent,Pods}` fields into the HPA spec
+- `HorizontalPodAutoscalerBehavior` should be added as a field to the HPA spec. It will contain fields
+  which describe the scaling behavior for both scale up and scale down. In the future other aspects of
+  the scaling behavior could be customized by adding fields here.
+- `HPAScaleConstraint` specifies the maximum change in the number of pods while scaling up or down
+  during autoscaling.
 
-The resulting solution will look like this:
+The resulting data structures will look like this:
 
 ```golang
-type HPAScaleConstraintValue struct {
-    Rate         *HPAScaleConstraintRateValue
-    DelaySeconds *int32
-
-type HPAScaleConstraintRateValue struct {
-    Pods          *int32
-    Percent       *int32
-    PeriodSeconds *int32
+type HPAScalingPolicyType string
+
+const (
+  PercentPolicy HPAScalingPolicyType = "percent"
+  PodsPolicy    HPAScalingPolicyType = "pods"
+)
+
+type HPAScalingPolicy struct {
+  Type          HPAScalingPolicyType
+  Value         int32
+  PeriodSeconds int32
 }
 
-type HPAScaleConstraints struct {
-    ScaleUp   *HPAScaleConstraintValue
-    ScaleDown *HPAScaleConstraintValue
+type HPAScalingRules struct {
+  Policies     []HpaScalingPolicy
+  SelectPolicy *string
+}
+
+type HorizontalPodAutoscalerBehavior struct {
```

Comment:
The word "Behavior" is very generic.  Can you say more about why you prefer this over "Constraint"?

Reply:
The way I see it there is the `metrics` section which the controller uses as inputs to scale the target. So we have the inputs covered but we have no means to controlling the output i.e. the autoscaling behavior. So this is the umbrella object to specify the the actions which the controller should perform. For instance the `stabilizationWindowSeconds` is a field to control how the autoscaling behaves and is not a constraint. I think there may be other aspects to the autoscaling behavior like the tolerance which should also be customizable(outside the scope of this KEP). But I'm open to suggestions.

Reply:
I'd assign names here the following way:
- `behavior` - is how the `desiredReplicas` is calculated
- `constraint` - is how the `desiredReplicas` is limited after that
So, I don't think that `behavior` is a proper name here.

Reply:
That's an internal detail which I think users should not have to know about.